### PR TITLE
Fabricate rewrite PR

### PR DESCRIPTION
**Pull request by aaronrudkin**
This is the PR that will merge in all of the new level syntax, including the cross_level stuff.

**Travis:**
Travis is currently failing on a specific test even though R CMD Check on my local system is working. @nfultz figures it's a weird Travis quirk, but it's POSSIBLE that there is an actual build issue (see builds 280, 281, 283). I'll be back this evening if I'm needed to address this stuff.

**Notes about backwards compatibility:**
The one thing that I am not 100% certain about in terms of replicating the existing behaviour of the old level function is the exact conditions under which a new ID label column is added to data. This is not breaking when it comes to any substantive stuff, it's just that a data frame might have an extra ID column relative to some past workflows. I began a discussion with @graemeblair about this in issue #33 -- I just haven't fully doubled back and verified exactly what's going on in the code.

**Not done:**
Vignettes related to cross-classifying data; draw_discrete changes

**Code review / refactoring notes:**
Some of the error handling and setup for the level functions is slightly duplicated and could possibly be abstracted into helper functions; the fabricate.r file should be split up into multiple files.

### CODE REVIEW

File: R/cross_classify_helpers.R [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  # Error handling for N
+  if(is.null(N) || is.na(N) || !is.atomic(N) || length(N) > 1 || N <= 0) {
+    stop("N must be a single integer that is positive.")
+  }
+
+  # Error handling for rho, if specified
+  if(is.null(sigma)) {
+    if(is.atomic(rho) & length(rho)==1) {
+      if(ndim>2 & rho<0) {
+        stop("The correlation matrix must be positive semi-definite. In specific, ",
```
Comment (nfultz) [inline R/cross_classify_helpers.R]: "Specifically"

File: R/cross_classify_helpers.R [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+  # Error handling for sigma
+  if(ncol(sigma) != ndim | nrow(sigma) != ndim | any(diag(sigma) != 1)) {
+    stop("The correlation matrix must be square, with the number of dimensions ",
+         "equal to the number of dimensions you are drawing from. In addition, ",
+         "the diagonal of the matrix must be equal to 1.")
+  }
+
+  # Can we use the fast package or are we stuck with the slow one?
+  use_f = use_f && requireNamespace("mvnfast", quietly = TRUE)
```
Comment (nfultz) [inline R/cross_classify_helpers.R]: :+1:

File: R/cross_classify_helpers.R [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    right_chol = right_chol[, order(attr(right_chol, "pivot"))]
+    # Generate standard normal data and right-multiply by decomposed matrix
+    # with right_chol to make it correlated.
+    correlated_sn <- matrix(rnorm(N * ndim),
+                            nrow = N,
+                            byrow = TRUE) %*% right_chol
+
+  } else {
+    # Using mvnfast
+    correlated_sn = mvnfast::rmvn(N, ncores = 2, mu, sigma)
```
Comment (nfultz) [inline R/cross_classify_helpers.R]: `getOption("mc.cores")`

File: R/cross_classify_helpers.R [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    # Right cholesky decomposition (i.e. LR = sigma s.t. L is lower triang, R
+    # is upper triang.)
+    right_chol = chol(sigma, pivot=TRUE)
+    # Order columns by the pivot attribute -- induces numerical stability?
+    right_chol = right_chol[, order(attr(right_chol, "pivot"))]
+    # Generate standard normal data and right-multiply by decomposed matrix
+    # with right_chol to make it correlated.
+    correlated_sn <- matrix(rnorm(N * ndim),
+                            nrow = N,
+                            byrow = TRUE) %*% right_chol
```
Comment (nfultz) [inline R/cross_classify_helpers.R]: No need for byrow

File: R/draw_binary_icc.R [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#' the supplied cluster IDs.
+#' @examples
+#' clusters = rep(1:5, 10)
+#' draw_binary_icc(clusters = clusters)
+#' draw_binary_icc(x = 0.5, clusters = clusters, rho = 0.5)
+#'
+#' @importFrom stats rbinom
+#'
+#' @export
+draw_binary_icc = function(x = 0.5, N = NULL, clusters, rho = 0.5) {
```
Comment (nfultz) [inline R/draw_binary_icc.R]: Mild preference for 0 or 1 to be the default rho here and for normal.

File: R/helper_functions.R [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+  if(is.symbol(l_arg)) {
+    # If it's a symbol, return the symbol
+    return(unname(l_arg))
+  } else if(is.language(l_arg)) {
+    # If it's a language call, then we need to unpack some more
+    # Extract the language from the language call
+    recurse = lang_args(l_arg)
+    # Iterate through each part of the language, recursively calling this function
+    # Results are a list, so unlist and unname to flatten
+    temp = unname(unlist(lapply(recurse, function(i) { get_symbols_from_expression(i) })))
```
Comment (nfultz) [inline R/helper_functions.R]: I've been using rapply for this, your might be better.

File: R/fabricate.R [hunk 7]
```diff
@@ -35,215 +52,667 @@
...
+    stop(
+      "Fabricate can be called in one of three ways: \n",
+      "1) Provide one or more level calls, with or without existing data \n",
+      "2) Provide existing data and add new variables without adding a level \n",
+      "3) Provide an \"N\" and add new variables"
+      )
+  }
+
+  # Create a blank working environment.
+  working_environment = new.env()
```
Comment (nfultz) [inline R/fabricate.R]: also `parent=emptyenv()` is a bit more defensive - see eg my ht package.

File: R/fabricate.R [hunk 8]
```diff
@@ -35,215 +52,667 @@
...
+  # or a series of level calls. You can't mix and match.
+  # This helper function will be TRUE if calls are all levels, FALSE
+  # if there are no calls or they are not levels.
+  all_levels = check_all_levels(data_arguments)
+
+  # User must provide exactly one of:
+  # 1) One or more level calls (with or without importing their own data)
+  # 2) Import their own data and do not involve level calls
+  # 3) Provide an N without importing their own data
+  if(sum((!is.null(data) & !missing(data) & !all_levels),
```
Comment (nfultz) [inline R/fabricate.R]: missing should be to the left of the is.null, also use &&s here to short circuit it. Also the comment is better than the error message below.

File: R/fabricate.R [hunk 9]
```diff
@@ -35,215 +52,667 @@
...
-      )
-    } else if(is.numeric(N) & any(!N%%1 == 0)) {
-      stop(
-        "The provided N must be an integer number. Provided N was of type ",
-        typeof(N)
-      )
+  # Pass-through mapper to nest_level.
+  # This needs to be done after we read the working environment and
+  # before we check N or do the shelving procedure.
+  if(nest &
```
Comment (nfultz) [inline R/fabricate.R]: &&

File: R/fabricate.R [hunk 10]
```diff
@@ -35,215 +52,667 @@
...
+                      working_environment_=working_environment_,
+                      data_arguments=data_arguments))
+  }
+
+  # Check to make sure the N here is sane
+  N = handle_n(N, add_level=TRUE)
+
+  # User is adding a new level, but already has a working data frame.
+  # Shelf the working data frame and move on
+  if("data_frame_output_" %in% names(working_environment_)) {
```
Comment (nfultz) [inline R/fabricate.R]: This logic is a bit hard to follow, going to meditate on it.

File: R/fabricate.R [hunk 11]
```diff
@@ -35,215 +52,667 @@
...
+  unique_values_of_level = unique(working_environment_$data_frame_output_[[ID_label]])
+
+  # Pre-allocate the mapping vector
+  index_maps = numeric(length(working_environment_$data_frame_output_[[ID_label]]))
+  # Iterate along the unique values of the level
+  for(i in seq_along(unique_values_of_level)) {
+    # Any obs that matches the level matching this i will be a duplicate of this i.
+    index_maps[
+      working_environment_$data_frame_output_[[ID_label]] == unique_values_of_level[i]
+      ] = i
```
Comment (nfultz) [inline R/fabricate.R]: yikes.

File: R/fabricate.R [hunk 12]
```diff
@@ -35,215 +52,667 @@
...
+  }
+
+  if(is.null(by) || !length(by$variable_names)) {
+    stop("You must specify a join structure to create cross-classified data.")
+  }
+
+  # Move the current working data frame into a package
+  package_df = list(data_frame_output_ = working_environment_$data_frame_output_,
+                    level_ids_ = working_environment_$level_ids_,
+                    variable_names_ = names(working_environment_$data_frame_output_))
```
Comment (nfultz) [inline R/fabricate.R]: this snippet pops up a bit, might want to make a helper / constructor function.

File: R/fabricate.R [hunk 13]
```diff
@@ -35,215 +52,667 @@
...
+
+  # Clear the active working data frame.
+  working_environment_$data_frame_output_ =
+    working_environment_$level_ids_ =
+    working_environment_$variable_names_ = NULL
+
+  # Loop over the variable name
+
+  variable_names = by$variable_names
+  data_frame_indices = numeric(length(variable_names))
```
Comment (nfultz) [inline R/fabricate.R]: ?

File: R/cross_classify_helpers.R [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+  colnames(quantiles) = names(data_list)
+
+  # Quantiles to inverse eCDF.
+  result = lapply(
+    seq_along(data_list),
+    function(vn) {
+      # What would the indices of the quantiles be if our data was ordered --
+      # if the answer is below 0, set it to 1. round will ensure the tie-
+      # breaking behaviour is random with respect to outcomes
+      ordered_indices = pmax(1,
```
Comment (nfultz) [inline R/cross_classify_helpers.R]: this can be factored out of the lapply to a seperate sweep statement -

File: R/fabricate.R (hunk 13 above)
Comment (nfultz) [reply R/fabricate.R]: length always returns an int I believe.

File: tests/testthat/test-crossclassified.R [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+  expect_error(fabricatr:::join_dfs(list(df1, df2), c("j1", "j2"),
+                                   N=100,
+                                   sigma=matrix(c(1, 0.3, 0.3, 0.3, 1, 0.3, 0.3, 0.3, 1),
+                                                ncol = 3
+                                                )))
+})
+
+test_that("Deliberate failures in cross_level", {
+  expect_error(
+    test_next = fabricate(
```
Comment (nfultz) [inline tests/testthat/test-crossclassified.R]: you should not have a name (test_next) on this expect_error (test-crossclassified line 96) -

File: R/cross_classify_helpers.R (hunk 1 above)
Comment (aaronrudkin) [reply R/cross_classify_helpers.R]: omg grammar correction in a code review

File: R/cross_classify_helpers.R (hunk 3 above)
Comment (aaronrudkin) [reply R/cross_classify_helpers.R]: oh, nuts, this is clearly an artifact from putzing around on my machine. Will fix.

File: R/fabricate.R (hunk 11 above)
Comment (aaronrudkin) [reply R/fabricate.R]: Yeah I'm really open to a better implementation of this, I'm sure there is one.

File: R/draw_binary_icc.R (hunk 5 above)
Comment (aaronrudkin) [reply R/draw_binary_icc.R]: Agreed.

**Comment by aaronrudkin** [comment]
I believe I've implemented all of @nfultz's suggestions that are trivial or blocking, the remainder I've issued. Provided everyone else is fine, this can be squashed and merged now. Phew. Feels good to have over a month of hard work finally be merged into master.
